File: 5.01.00.2163/src/engines_gpl/waq/packages/waq_io/src/ods/julian.f
!!  Copyright(C) Stichting Deltares, 2012.
!!
!!  This program is free software: you can redistribute it and/or modify
!!  it under the terms of the GNU General Public License version 3,
!!  as published by the Free Software Foundation.
!!
!!  This program is distributed in the hope that it will be useful,
!!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
!!  GNU General Public License for more details.
!!
!!  You should have received a copy of the GNU General Public License
!!  along with this program. If not, see <http://www.gnu.org/licenses/>.
!!
!!  contact: delft3d.support@deltares.nl
!!  Stichting Deltares
!!  P.O. Box 177
!!  2600 MH Delft, The Netherlands
!!
!!  All indications and logos of, and references to registered trademarks
!!  of Stichting Deltares remain the property of Stichting Deltares. All
!!  rights reserved.

      DOUBLE PRECISION FUNCTION JULIAN ( IDATE , ITIME )
C
C     +----------------------------------------------------------------+
C     |    W A T E R L O O P K U N D I G   L A B O R A T O R I U M     |
C     |               Sector Waterbeheer & Milieu                      |
C     +----------------------------------------------------------------+
C
C***********************************************************************
C
C     Project : T0467
C     Author  : Andre Hendriks
C     Date    : 891215             Version : 1.00
C
C     Changes in this module :
C
C     Date    Author          Description
C     ------  --------------  -----------------------------------
C     ......  ..............  ..............................
C     891215  Andre Hendriks  Version 1.00
C
C***********************************************************************
C
C     Description of module :
C
C        This functions returns the so called Julian day of a date, or
C        the value -1.0 if an error occurred.
C
C        The Julian day of a date is the number of days that has passed
C        since January 1, 4712 BC at 12h00 ( Gregorian). It is usefull
C        to compute differces between dates. ( See SUBROUTINE GREGOR
C        for the reverse proces ).
C
C***********************************************************************
C
C     Arguments :
C
C     Name   Type     In/Out Size            Description
C     ------ -----    ------ -------         ---------------------------
C     IDATE  integer  in     -               Date as YYYYMMDD
C     ITIME  integer  in     -               Time as HHMMSS
C
C     Local variables :
C
C     Name   Type     Size   Description
C     ------ -----    ------ ------------------------
C     TEMP1  real*8   -      Temporary variable
C     TEMP2  real*8   -      Temporary variable
C     IYEAR  integer  -      Year   ( -4713-.. )
C     IMONTH integer  -      Month  ( 1-12 )
C     IDAY   integer  -      Day    ( 1-28,29,30 or 31 )
C     IHOUR  integer  -      Hour   ( 0-23 )
C     IMIN   integer  -      Minute ( 0-59 )
C     ISEC   integer  -      Second ( 0-59 )
C     MONLEN integer  12     Length of month in days
C
C     Calls to : none
C
C***********************************************************************
C
C     Variables :
C
      INTEGER          IYEAR , IMONTH, IDAY  , IHOUR , IMIN  , ISEC  ,
     1                 IDATE , ITIME , MONLEN(12)
      DOUBLE PRECISION TEMP1 , TEMP2
      CHARACTER*48     LINE
C
C***********************************************************************
C
C     Initialize lenghts of months :
C
      DATA MONLEN / 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 /
C
C***********************************************************************
C
C
C
      IYEAR  = IDATE/10000
      IMONTH = IDATE/100 - IYEAR*100
      IDAY   = IDATE - IYEAR*10000 - IMONTH*100
      IHOUR  = ITIME/10000
      IMIN   = ITIME/100 - IHOUR*100
      ISEC   = ITIME - IHOUR*10000 - IMIN*100
      IF (( IYEAR  .LT. -4713 ) .OR. ( IMONTH .LT.  1 ) .OR.
     1    ( IMONTH .GT.    12 ) .OR. ( IDAY   .LT.  1 ) .OR.
     2    ( IDAY   .GT. MONLEN(IMONTH) ) .OR.
     3    ( IHOUR  .LT.     0 ) .OR. ( IHOUR  .GT. 24 ) .OR.
     4    ( IMIN   .LT.     0 ) .OR. ( IMIN   .GT. 60 ) .OR.
     5    ( ISEC   .LT.     0 ) .OR. ( ISEC   .GT. 60 )) THEN
         JULIAN = -1.0
         WRITE(LINE,'(A33,I8,''-'',I6)') 'ERROR in JULIAN interpeting time:',IDATE,ITIME
         CALL MONSYS(LINE,1)
         GOTO 999
      ELSE
         TEMP1  = INT (( IMONTH-14.0) / 12.0 )
*-----------------------------
c        WRITE(*,*) 'TEMP1 : ', TEMP1
c        WRITE(*,*) 'dbg 1 : ',
c    1          INT ( 1461.0 * ( IYEAR + 4800.0 + TEMP1 ) / 4.0 )
c        WRITE(*,*) 'dbg 2 : ',
c    2          INT ( 367.0 * ( IMONTH - 2.0 - TEMP1 * 12.0 ) / 12.0 )
c        WRITE(*,*) 'dbg 3 : ',
c    1          INT ( 3.0 * INT ( ( IYEAR + 4900.0 + TEMP1 ) / 100.0 ) /
c    2          4.0 )
*-----------------------------
         TEMP2  = IDAY - 32075.0 +
     1          INT ( 1461.0 * ( IYEAR + 4800.0 + TEMP1 ) / 4.0 ) +
     2          INT ( 367.0 * ( IMONTH - 2.0 - TEMP1 * 12.0 ) / 12.0 ) -
     3          INT ( 3.0 * INT ( ( IYEAR + 4900.0 + TEMP1 ) / 100.0 ) /
     4          4.0 )
         TEMP1  = FLOAT ( IHOUR ) * 3600.0 +
     1            FLOAT ( IMIN  ) *   60.0 +
     2            FLOAT ( ISEC  ) - 43200.0
         JULIAN = TEMP2 + ( TEMP1 / 86400.0 )
      ENDIF
  999 RETURN
      END

File: 5.01.00.2163/src/utils_lgpl/ods/packages/ods/src/julian.c
/*
 *  julian.c  -  ODS convert Gregorian to Julian date
 *
 *  Copyright (C) 1994 Delft Hydraulics
 *
 *  Marc Kool
 */

/*
 *  $Author: Markus $
 *  $Date: 12/27/00 2:53p $
 *  $Source: /u/cvsroot/gpp/libsrc/ods/julian.c,v $
 *
 */


/*   Date:       12 Apr 1994                                          */
/*   Time:       15:55                                                */
/*   Program:    JULIAN.C                                             */
/*   Version:    1.02                                                 */
/*   Programmer: Andr‚ Hendriks                                       */
/*   (c) Copyright 1993 Delft Hydraulics                              */
/*   Previous version(s):                                             */
/*   1.01 -- 9 Dec 1993 -- 13:00 -- Operating System: DOS            */
/*   1.00 -- 27 Aug 1993 -- 19:23 -- Operating System: DOS            */
/*   0.00 -- 4 Aug 1993 -- 08:21 -- Operating System: DOS             */
/*   Project:    Open Data Structuur                                  */
/*   Module:     Julian                                               */
/*   Function:   This functions returns the so called Julian day of   */
/*               a date, or the value -1.0 if an error occurred.      */
/*               The Julian day of a date is the number of days that  */
/*               has passed since January 1, 4712 BC at 12h00         */
/*               ( Gregorian). It is usefull to compute differces     */
/*               between dates. ( See function gregot for the         */
/*               reverse proces ).                                    */
/*   Comment:                                                         */
/*   Reference:                                                       */
/*   Review:                                                          */

#include "portable.h"
#include "ods.h"

#include <math.h>

#undef SUN
#ifdef SUN
#   define JULIAN julian_
#else
#   define JULIAN julian
#endif

#ifndef NOPROT
   void FUNTYPE JULIAN ( TInt4 *iyear ,
                                   TInt4 *imonth,
                                   TInt4 *iday  ,
                                   TInt4 *ihour ,
                                   TInt4 *imin  ,
                                   TInt4 *isec  ,
                                   TReal8 *jdate)
#else
   void FUNTYPE JULIAN ( iyear ,
                                   imonth,
                                   iday  ,
                                   ihour ,
                                   imin  ,
                                   isec  ,
                                   jdate)
   TReal8 *jdate ;
   TInt4 *iyear, *imonth, *iday, *ihour, *imin, *isec ;
#endif

/*************************************************************************/
/*                                                                       */
/*     Arguments:                                                        */
/*                                                                       */
/*        Name   In/Out Size            Description                      */
/*        ------ ------ -------         ---------------------------      */
/*        jdate  out    -               Julian day                       */
/*        iyear  in     -               Year   ( -4713-.. )              */
/*        imonth in     -               Month  ( 1-12 )                  */
/*        iday   in     -               Day    ( 1-28,29,30 or 31 )      */
/*        ihour  in     -               Hour   ( 0-23 )                  */
/*        imin   in     -               Minute ( 0-59 )                  */
/*        isec   in     -               Second ( 0-59 )                  */
/*                                                                       */
/*************************************************************************/

   {
   TReal8 temp1, temp2 ;
   TInt4    monlen [] = { 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31} ;

   if (( *iyear  < -4713 ) || ( *imonth <  1 ) ||
       ( *imonth >    12 ) || ( *iday   <  1 ) ||
       ( *iday   > monlen[ *imonth -1 ]) ||
       ( *ihour  <     0 ) || ( *ihour  > 24 ) ||
       ( *imin   <     0 ) || ( *imin   > 60 ) ||
       ( *isec   <     0 ) || ( *isec   > 60 ))
      *jdate  = -1.0 ;
   else
      {
      if (( *imonth == 1) || ( *imonth == 2))
         temp1 = -1.0 ;
      else
         temp1 =  0.0 ;
      temp2  = *iday - 32075.0 +
               (TInt4) ( 1461.0 * ( *iyear + 4800.0 + temp1) / 4.0) +
               (TInt4) ( 367 * ( *imonth - 2 - (TInt4) temp1 * 12) / 12) -
               (TInt4) ( 3.0 * (TInt4) (( *iyear + 4900.0 + temp1) / 100.0) / 4.0) ;
      temp1  = *ihour * 3600.0 + *imin * 60.0 + *isec - 43200.0 ;
      *jdate = temp2 + ( temp1 / 86400.0 ) ;
      }
   return ;
   }
